server/sonar-server/src/main/java/org/sonar/server/authentication/GenerateJwtTokenFilter.java

```diff
@@ -0,0 +1,142 @@
+import javax.servlet.ServletResponse;
+import javax.servlet.WriteListener;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpServletResponseWrapper;
+import org.sonar.api.web.ServletFilter;
+import org.sonar.server.user.UserSession;
+
+import static org.jboss.netty.handler.codec.http.HttpMethod.POST;
+
+/**
+ * This filter should only be executed on "/sessions/login".
+ *
+ * It will create a new session when user is authenticated.
+ */
+public class GenerateJwtTokenFilter extends ServletFilter {
+
+  private final JwtTokenUpdater jwtTokenUpdater;
+  private final UserSession userSession;
+
+  public GenerateJwtTokenFilter(JwtTokenUpdater jwtTokenUpdater, UserSession userSession) {
+    this.jwtTokenUpdater = jwtTokenUpdater;
+    this.userSession = userSession;
+  }
+
+  @Override
+  public UrlPattern doGetPattern() {
+    return UrlPattern.create("/sessions/login");
+  }
+
+  @Override
+  public void doFilter(ServletRequest servletRequest, ServletResponse servletResponse, FilterChain chain) throws IOException, ServletException {
+    HttpServletRequest request = (HttpServletRequest) servletRequest;
+    HttpServletResponse response = (HttpServletResponse) servletResponse;
+
+    if (request.getMethod().equals(POST.getName())) {
+      BufferResponseWrapper wrapper = new BufferResponseWrapper(response);
+      chain.doFilter(request, wrapper);
+      if (userSession.isLoggedIn()) {
+        jwtTokenUpdater.createNewJwtToken(userSession.getLogin(), response);
+      }
+      response.getOutputStream().write(wrapper.getWrapperBytes());
+    } else {
+      chain.doFilter(request, response);
+    }
+  }
+
+  @Override
+  public void init(FilterConfig filterConfig) throws ServletException {
+    // Nothing to do
+  }
+
+  @Override
+  public void destroy() {
+    // Nothing to do
+  }
+
+  /**
+   * As the RackFilter is executed before this filter, the reponse is commited and it's not possible anymore to add cookie.
+   * So we're create a buffer response wrapper that will buffer the dat that should be send to the browser in order to not commit the response.
+   * It's then possible to add cookie before flushing data to the browser.
+   *
+   * See <a href="http://stackoverflow.com/questions/11025605/response-is-committing-and-dofilter-chain-is-broken">
+   *
+   * Note : this must be removed when authentication will not use rails anymore
+   */
+  private static final class BufferResponseWrapper extends HttpServletResponseWrapper {
+
+    private BufferServletOutputStream stream = new BufferServletOutputStream();
+
+    BufferResponseWrapper(HttpServletResponse httpServletResponse) {
+      super(httpServletResponse);
+    }
+
+    @Override
+    public ServletOutputStream getOutputStream() throws IOException {
+      return stream;
+    }
+
+    @Override
+    public PrintWriter getWriter() throws IOException {
+      return new PrintWriter(stream);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this use of constructor "PrintWriter(OutputStream)" [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1943)